### PR TITLE
Wdt automate tokenize credentials

### PR DESCRIPTION
**Pull request by CarolynRountree**
discoverDomain changed to automatically replace password attribute values with a property token, and the property written a specified variable file IF the -variable_file argument is included on the command line. 

This will work in conjunction with any variable injector directives the user specifies.

### CODE REVIEW

File: installer/src/main/bin/discoverDomain.sh [hunk 1]
```diff
@@ -71,6 +103,12 @@ usage() {
   echo "        model_file      - the location of the model file to use,"
   echo "                          the default is to get the model from the archive"
   echo ""
+  echo "        variable_file    - the location of the variable file to write ,"
```
Comment (rakillen) [inline installer/src/main/bin/discoverDomain.sh]: The extra commas at the end are showing up in the script output.
The hyphen and text is indented one extra space.

Comment (CarolynRountree) [reply installer/src/main/bin/discoverDomain.sh]: fixed thank you

File: core/src/main/resources/oracle/weblogic/deploy/messages/wlsdeploy_rb.properties [hunk 2]
```diff
@@ -472,6 +472,10 @@ WLSDPLY-06021=The variables file command line argument {0} was used but the mode
   {1} does not exist : {2}
 WLSDPLY-06022=Discover domain {0}
 WLSDPLY-06023=No domain name found in the domain configuration
+WLSDPLY-06024=Variable file not provided as a command line argument. Automatic replacement of passwords with tokens \
```
Comment (ddsharpe) [inline core/src/main/resources/oracle/weblogic/deploy/messages/wlsdeploy_rb.properties]: I think we should brainstorm a few minutes on these messages.  6025 could be a little clearer.

Comment (CarolynRountree) [reply core/src/main/resources/oracle/weblogic/deploy/messages/wlsdeploy_rb.properties]: done

File: site/discover.md [hunk 3]
```diff
@@ -1,12 +1,66 @@
...
 
-To run the Discover Domain Tool, simply provide the domain location and the name of the archive file; a separate model file can also be provided to make editing the generated model easier.  For example:
+To run the Discover Domain Tool, simply provide the oracle home and domain location. Provide a location and file name for the model file and archive file where the discovered domain information will be written. The model can be generated in either yaml or json format. Simply label the file name with the correct suffix. Both a sh and cmd script are provided.
```
Comment (rakillen) [inline site/discover.md]: *Oracle, YAML, JSON

File: site/discover.md [hunk 4]
```diff
@@ -1,12 +1,66 @@
...
+    
+    weblogic-deploy\bin\discoverDomain.cmd -oracle_home c:\wls12213 -domain_home domains\DemoDomain -archive_file DiscoveredDemoDomain.zip -model_file DiscoveredDemoDomain.yaml -domain_type RestrictedJRF
+    
+Before the model is persisted to the model file, any variable injectors or model filters are run, in that order. The final step is validation, which validates the contents of the model, archive and variable file. If the validation is successful, the model is persisted. For more information on these three topics, go to:
+
+ - [Variable Injection](site/variable_injection.md)
+ - [Model Filters](site/tool_filters.md)
+ - [Validate Model Tool](site/validate.md)
+
+Any problems (or success) will be listed in the discover tool summary. The summary will print the version of the tool and oracle home, and the WLST mode with which the tool was run (online or offline). A recap of all Warning and Severe messages will be listed, along with a total.
```
Comment (rakillen) [inline site/discover.md]: *Oracle

File: site/discover.md (hunk 3 above)
Comment (rakillen) [inline site/discover.md]: *Oracle

File: site/discover.md (hunk 4 above)
Comment (rakillen) [inline site/discover.md]: *Oracle

File: site/discover.md (hunk 3 above)
Comment (CarolynRountree) [reply site/discover.md]: done.

File: core/src/main/python/discover.py [hunk 5]
```diff
@@ -189,20 +189,20 @@ def __process_variable_filename_arg(optional_arg_map):
     """
     _method_name = '__process_variable_filename_arg'
 
-    if CommandLineArgUtil.VARIABLE_PROPERTIES_FILE_SWITCH in optional_arg_map:
+    if CommandLineArgUtil.VARIABLE_FILE_SWITCH in optional_arg_map:
         variable_injector_file_name = variable_injector.get_default_variable_injector_file_name()
```
Comment (ddsharpe) [inline core/src/main/python/discover.py]: the variable file is always the default, and never the name of the file provided on the command line switch

**Comment by ddsharpe** [comment]
the default variable file, model.properties, is zeroed out even if I do not put the -variable_file argument on the command line

**Comment by CarolynRountree** [comment]
> 
> 
> These properties for the datasources are really long. Is it hard to shorten them?
> Currently:
> JDBCSystemResource.Derek-DataSource.JdbcResource.JDBCDriverParams.PasswordEncrypted=
> 
> Suggested alternatives:
> Derek-DataSource.PasswordEncrypted=
> Datasource.Derek-DataSource.PasswordEncrypted=
> JDBCSystemResource.Derek-DataSource.PasswordEncrypted=

One algorithm I can think of is to replace every MBean/Name pair, with Name, and remove each singleton MBean, except for the last one. So that the long name above becomes
Derek-DataSource.JDBCDriverParams.PasswordEncrypted - Richard, Derek any more ideas on how to get it shorter with an automated algorithm?

File: core/src/main/python/wlsdeploy/tool/discover/discoverer.py [hunk 6]
```diff
@@ -677,6 +682,29 @@ def _inspect_security_folder_name(self, folder_name, location):
 
         return folder_name
 
+    def _inject_token(self, model_section, attribute, location, injector_commands=OrderedDict()):
```
Comment (rakillen) [inline core/src/main/python/wlsdeploy/tool/discover/discoverer.py]: not sure why this was added.  a strange test even if it was un-commented

Comment (rakillen) [reply core/src/main/python/wlsdeploy/tool/discover/discoverer.py]: ok now

File: core/src/test/python/collections_test.py [hunk 7]
```diff
@@ -197,5 +197,30 @@ def testDictOrderedDictUpdate(self):
         self.assertEqual('entry1' in dict1 and 'entry2' in dict1, True,
                          "expected ordereddict1 to contain 'entry1' and 'entry2' keys")
 
+    # def testOrderedDictSort(self):
```
Comment (rakillen) [inline core/src/test/python/collections_test.py]: Why did DS add this?

File: core/src/main/python/wlsdeploy/tool/util/variable_injector.py [hunk 8]
```diff
@@ -261,8 +335,21 @@ def _traverse_variables(model_section, mbean_list):
 
         return variable_dict
 
-    def __format_variable_name(self, location, attribute):
-        _method_name = '__format_variable_name'
+    def get_folder_short_name(self, folder_name):
+        _method_name = '_match_short_top_folder'
```
Comment (rakillen) [inline core/src/main/python/wlsdeploy/tool/util/variable_injector.py]: Method name is 'get_folder_short_name'

File: core/src/test/python/collections_test.py (hunk 7 above)
Comment (CarolynRountree) [reply core/src/test/python/collections_test.py]: I thought PyOrderedDict had a problem. This test exposed an issue with my own logic, so I asked him to add it permanently. It's a simple to test to sort a dictionary by keys, and add the items to the dictionary in this sorted order. The PyOrderedDict simply cares about the order you add them, but its a good example of how to sort a dictionary with our OrderedDict

**Pull request merged**